In [2]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn import preprocessing


We import all the various datasets

In [3]:
mcf7_smartS_meta = pd.read_csv("Data/SmartSeq/MCF7_SmartS_MetaData.tsv",delimiter="\t", index_col=0)
mcf7_smartS_filn = pd.read_csv("Data/SmartSeq/MCF7_SmartS_Filtered_Normalised_3000_Data_train.txt",delimiter=" ",index_col=0).T
hcc_smartS_meta = pd.read_csv("Data/SmartSeq/HCC1806_SmartS_MetaData.tsv",delimiter="\t",index_col=0)
hcc_smartS_filn = pd.read_csv("Data/SmartSeq/HCC1806_SmartS_Filtered_Normalised_3000_Data_train.txt",delimiter=" ",index_col=0).T
mcf7_dropS_filn = pd.read_csv("Data/DropSeq/MCF7_Filtered_Normalised_3000_Data_train.txt",delimiter=" ",index_col=0).T
hcc_dropS_filn = pd.read_csv("Data/DropSeq/HCC1806_Filtered_Normalised_3000_Data_train.txt",delimiter=" ",index_col=0).T
mcf7_smartS_filt = pd.read_csv("Data/SmartSeq/MCF7_SmartS_Filtered_Data.txt",delimiter=" ",index_col=0).T
hcc_smartS_filt = pd.read_csv("Data/SmartSeq/HCC1806_SmartS_Filtered_Data.txt",delimiter=" ",index_col=0).T
mcf7_smartS_raw = pd.read_csv("Data/SmartSeq/MCF7_SmartS_Unfiltered_Data.txt",delimiter=" ",index_col=0).T
hcc_smartS_raw = pd.read_csv("Data/SmartSeq/HCC1806_SmartS_Unfiltered_Data.txt",delimiter=" ",index_col=0).T

Since we have unfiltered data only for the Smart sequencing, we will focus on Smartseq. Let us start with MCF7

In [4]:
mcf7_smartS_raw

WASH7P  MIR6859-1  WASH9P  \
output.STAR.1_A10_Hypo_S28_Aligned.sortedByCoor...       0          0       1   
output.STAR.1_A11_Hypo_S29_Aligned.sortedByCoor...       0          0       0   
output.STAR.1_A12_Hypo_S30_Aligned.sortedByCoor...       0          0       0   
output.STAR.1_A1_Norm_S1_Aligned.sortedByCoord....       0          0       0   
output.STAR.1_A2_Norm_S2_Aligned.sortedByCoord....       0          0       0   
...                                                    ...        ...     ...   
output.STAR.4_H5_Norm_S359_Aligned.sortedByCoor...       0          0       0   
output.STAR.4_H6_Norm_S360_Aligned.sortedByCoor...       0          0       1   
output.STAR.4_H7_Hypo_S379_Aligned.sortedByCoor...       1          0       1   
output.STAR.4_H8_Hypo_S380_Aligned.sortedByCoor...       0          0       4   
output.STAR.4_H9_Hypo_S381_Aligned.sortedByCoor...       1          0       5   

                                                    OR4F29  MTND1P23  \
output.STAR.1_A10_Hypo_S28_Aligned.sortedByCoor...       0         0   
output.STAR.1_A11_Hypo_S29_Aligned.sortedByCoor...       0         0   
output.STAR.1_A12_Hypo_S30_Aligned.sortedByCoor...       0         0   
output.STAR.1_A1_Norm_S1_Aligned.sortedByCoord....       0         0   
output.STAR.1_A2_Norm_S2_Aligned.sortedByCoord....       0         0   
...                                                    ...       ...   
output.STAR.4_H5_Norm_S359_Aligned.sortedByCoor...       0         0   
output.STAR.4_H6_Norm_S360_Aligned.sortedByCoor...       0         0   
output.STAR.4_H7_Hypo_S379_Aligned.sortedByCoor...       0         0   
output.STAR.4_H8_Hypo_S380_Aligned.sortedByCoor...       1         0   
output.STAR.4_H9_Hypo_S381_Aligned.sortedByCoor...       0         0   

                                                    MTND2P28  MTCO1P12  \
output.STAR.1_A10_Hypo_S28_Aligned.sortedByCoor...         2         2   
output.STAR.1_A11_Hypo_S29_Aligned.sortedByCoor...         0         0   
output.STAR.1_A12_Hypo_S30_Aligned.sortedByCoor...         1         1   
output.STAR.1_A1_Norm_S1_Aligned.sortedByCoord....         0         0   
output.STAR.1_A2_Norm_S2_Aligned.sortedByCoord....         0         0   
...                                                      ...       ...   
output.STAR.4_H5_Norm_S359_Aligned.sortedByCoor...         0         1   
output.STAR.4_H6_Norm_S360_Aligned.sortedByCoor...         2         5   
output.STAR.4_H7_Hypo_S379_Aligned.sortedByCoor...         7         0   
output.STAR.4_H8_Hypo_S380_Aligned.sortedByCoor...        29         4   
output.STAR.4_H9_Hypo_S381_Aligned.sortedByCoor...         5         3   

                                                    MTCO2P12  MTATP8P1  \
output.STAR.1_A10_Hypo_S28_Aligned.sortedByCoor...         0         0   
output.STAR.1_A11_Hypo_S29_Aligned.sortedByCoor...         0         0   
output.STAR.1_A12_Hypo_S30_Aligned.sortedByCoor...         1         0   
output.STAR.1_A1_Norm_S1_Aligned.sortedByCoord....         0         0   
output.STAR.1_A2_Norm_S2_Aligned.sortedByCoord....         0         0   
...                                                      ...       ...   
output.STAR.4_H5_Norm_S359_Aligned.sortedByCoor...         0         0   
output.STAR.4_H6_Norm_S360_Aligned.sortedByCoor...         5         0   
output.STAR.4_H7_Hypo_S379_Aligned.sortedByCoor...         0         0   
output.STAR.4_H8_Hypo_S380_Aligned.sortedByCoor...         0         0   
output.STAR.4_H9_Hypo_S381_Aligned.sortedByCoor...         0         0   

                                                    MTATP6P1  ...  MT-TH  \
output.STAR.1_A10_Hypo_S28_Aligned.sortedByCoor...        29  ...      0   
output.STAR.1_A11_Hypo_S29_Aligned.sortedByCoor...         0  ...      1   
output.STAR.1_A12_Hypo_S30_Aligned.sortedByCoor...        12  ...      0   
output.STAR.1_A1_Norm_S1_Aligned.sortedByCoord....         7  ...      1   
output.STAR.1_A2_Norm_S2_Aligned.sortedByCoord....      

In [5]:
mcf7_smartS_filt

WASH7P  MIR6859-1  WASH9P  \
output.STAR.1_A10_Hypo_S28_Aligned.sortedByCoor...       0          0       1   
output.STAR.1_A12_Hypo_S30_Aligned.sortedByCoor...       0          0       0   
output.STAR.1_A2_Norm_S2_Aligned.sortedByCoord....       0          0       0   
output.STAR.1_A3_Norm_S3_Aligned.sortedByCoord....       0          0       1   
output.STAR.1_A4_Norm_S4_Aligned.sortedByCoord....       0          0      10   
...                                                    ...        ...     ...   
output.STAR.4_H5_Norm_S359_Aligned.sortedByCoor...       0          0       0   
output.STAR.4_H6_Norm_S360_Aligned.sortedByCoor...       0          0       1   
output.STAR.4_H7_Hypo_S379_Aligned.sortedByCoor...       1          0       1   
output.STAR.4_H8_Hypo_S380_Aligned.sortedByCoor...       0          0       4   
output.STAR.4_H9_Hypo_S381_Aligned.sortedByCoor...       1          0       5   

                                                    OR4F29  MTND1P23  \
output.STAR.1_A10_Hypo_S28_Aligned.sortedByCoor...       0         0   
output.STAR.1_A12_Hypo_S30_Aligned.sortedByCoor...       0         0   
output.STAR.1_A2_Norm_S2_Aligned.sortedByCoord....       0         0   
output.STAR.1_A3_Norm_S3_Aligned.sortedByCoord....       0         0   
output.STAR.1_A4_Norm_S4_Aligned.sortedByCoord....       0         0   
...                                                    ...       ...   
output.STAR.4_H5_Norm_S359_Aligned.sortedByCoor...       0         0   
output.STAR.4_H6_Norm_S360_Aligned.sortedByCoor...       0         0   
output.STAR.4_H7_Hypo_S379_Aligned.sortedByCoor...       0         0   
output.STAR.4_H8_Hypo_S380_Aligned.sortedByCoor...       1         0   
output.STAR.4_H9_Hypo_S381_Aligned.sortedByCoor...       0         0   

                                                    MTND2P28  MTCO1P12  \
output.STAR.1_A10_Hypo_S28_Aligned.sortedByCoor...         2         2   
output.STAR.1_A12_Hypo_S30_Aligned.sortedByCoor...         1         1   
output.STAR.1_A2_Norm_S2_Aligned.sortedByCoord....         0         0   
output.STAR.1_A3_Norm_S3_Aligned.sortedByCoord....         1         3   
output.STAR.1_A4_Norm_S4_Aligned.sortedByCoord....         3        10   
...                                                      ...       ...   
output.STAR.4_H5_Norm_S359_Aligned.sortedByCoor...         0         1   
output.STAR.4_H6_Norm_S360_Aligned.sortedByCoor...         2         5   
output.STAR.4_H7_Hypo_S379_Aligned.sortedByCoor...         7         0   
output.STAR.4_H8_Hypo_S380_Aligned.sortedByCoor...        29         4   
output.STAR.4_H9_Hypo_S381_Aligned.sortedByCoor...         5         3   

                                                    MTCO2P12  MTATP8P1  \
output.STAR.1_A10_Hypo_S28_Aligned.sortedByCoor...         0         0   
output.STAR.1_A12_Hypo_S30_Aligned.sortedByCoor...         1         0   
output.STAR.1_A2_Norm_S2_Aligned.sortedByCoord....         0         0   
output.STAR.1_A3_Norm_S3_Aligned.sortedByCoord....         0         0   
output.STAR.1_A4_Norm_S4_Aligned.sortedByCoord....         1         0   
...                                                      ...       ...   
output.STAR.4_H5_Norm_S359_Aligned.sortedByCoor...         0         0   
output.STAR.4_H6_Norm_S360_Aligned.sortedByCoor...         5         0   
output.STAR.4_H7_Hypo_S379_Aligned.sortedByCoor...         0         0   
output.STAR.4_H8_Hypo_S380_Aligned.sortedByCoor...         0         0   
output.STAR.4_H9_Hypo_S381_Aligned.sortedByCoor...         0         0   

                                                    MTATP6P1  ...  MT-TH  \
output.STAR.1_A10_Hypo_S28_Aligned.sortedByCoor...        29  ...      0   
output.STAR.1_A12_Hypo_S30_Aligned.sortedByCoor...        12  ...      0   
output.STAR.1_A2_Norm_S2_Aligned.sortedByCoord....        68  ...      0   
output.STAR.1_A3_Norm_S3_Aligned.sortedByCoord....       203  ...      1   
output.STAR.1_A4_Norm_S4_Aligned.sortedByCoord....      

In [6]:
mcf7_smartS_filn

CYP1B1  CYP1B1-AS1  \
output.STAR.2_B3_Norm_S57_Aligned.sortedByCoord...     343         140   
output.STAR.2_B4_Norm_S58_Aligned.sortedByCoord...     131          59   
output.STAR.2_B5_Norm_S59_Aligned.sortedByCoord...     452         203   
output.STAR.2_B6_Norm_S60_Aligned.sortedByCoord...      27           7   
output.STAR.2_B7_Hypo_S79_Aligned.sortedByCoord...    5817        2669   
...                                                    ...         ...   
output.STAR.4_H5_Norm_S359_Aligned.sortedByCoor...     196         102   
output.STAR.4_H6_Norm_S360_Aligned.sortedByCoor...     504         238   
output.STAR.4_H7_Hypo_S379_Aligned.sortedByCoor...   34565       13717   
output.STAR.4_H8_Hypo_S380_Aligned.sortedByCoor...   20024        7835   
output.STAR.4_H9_Hypo_S381_Aligned.sortedByCoor...    5953        2367   

                                                    CYP1A1  NDRG1  DDIT4  \
output.STAR.2_B3_Norm_S57_Aligned.sortedByCoord...       0      0    386   
output.STAR.2_B4_Norm_S58_Aligned.sortedByCoord...       0      1    289   
output.STAR.2_B5_Norm_S59_Aligned.sortedByCoord...       0      0      0   
output.STAR.2_B6_Norm_S60_Aligned.sortedByCoord...       0      0    288   
output.STAR.2_B7_Hypo_S79_Aligned.sortedByCoord...       0    654   2484   
...                                                    ...    ...    ...   
output.STAR.4_H5_Norm_S359_Aligned.sortedByCoor...       1    243    266   
output.STAR.4_H6_Norm_S360_Aligned.sortedByCoor...       0     62    417   
output.STAR.4_H7_Hypo_S379_Aligned.sortedByCoor...   11274   1263   4256   
output.STAR.4_H8_Hypo_S380_Aligned.sortedByCoor...     563    925  12733   
output.STAR.4_H9_Hypo_S381_Aligned.sortedByCoor...     522   1572   2275   

                                                    PFKFB3   HK2  AREG  MYBL2  \
output.STAR.2_B3_Norm_S57_Aligned.sortedByCoord...      75     0     0    476   
output.STAR.2_B4_Norm_S58_Aligned.sortedByCoord...      42     0   856   1586   
output.STAR.2_B5_Norm_S59_Aligned.sortedByCoord...       0     0     0    775   
output.STAR.2_B6_Norm_S60_Aligned.sortedByCoord...     214    15   242   1191   
output.STAR.2_B7_Hypo_S79_Aligned.sortedByCoord...    1603   748   436      0   
...                                                    ...   ...   ...    ...   
output.STAR.4_H5_Norm_S359_Aligned.sortedByCoor...     278    78     1    199   
output.STAR.4_H6_Norm_S360_Aligned.sortedByCoor...     932   275     0     33   
output.STAR.4_H7_Hypo_S379_Aligned.sortedByCoor...    5663   701     0      0   
output.STAR.4_H8_Hypo_S380_Aligned.sortedByCoor...    2187  1141     0     20   
output.STAR.4_H9_Hypo_S381_Aligned.sortedByCoor...    3269  2247   169      2   

                                                    ADM  ...  CD27-AS1  DNAI7  \
output.STAR.2_B3_Norm_S57_Aligned.sortedByCoord...    0  ...        63      0   
output.STAR.2_B4_Norm_S58_Aligned.sortedByCoord...    0  ...         0      0   
output.STAR.2_B5_Norm_S59_Aligned.sortedByCoord...    0  ...        15      0   
output.STAR.2_B6_Norm_S60_Aligned.sortedByCoord...    0  ...         1      0   
output.STAR.2_B7_Hypo_S79_Aligned.sortedByCoord...    0  ...         0      0   
...                                                 ...  ...       ...    ...   
output.STAR.4_H5_Norm_S359_Aligned.sortedByCoor...    0  ...        79      0   
output.STAR.4_H6_Norm_S360_Aligned.sortedByCoor...    0  ...        86      0   
output.STAR.4_H7_Hypo_S379_Aligned.sortedByCoor...    0  ...         0      0   
output.STAR.4_H8_Hypo_S380_Aligned.sortedByCoor...   81  ...         0      0   
output.STAR.4_H9_Hypo_S381_Aligned.sortedByCoor...    0  ...         0      0   

                                                    MAFG  LZTR1  BCO2  GRIK5  \
output.STAR.2_B3_Norm_S57_Aligned.sortedByCoord...    17     59     0      0   
output.STAR.2_B4_Norm_S58_Aligned.sortedByCoord...     1     25     0      0   
output.STAR.2_B5_Norm_S59_Aligned.sortedByCoord...    23      0 

We first check that the filtered dataset is a subset of the raw dataset, we do this by checking that the shape of the filtered dataset and the intersection between the latter and the raw one coincide.

In [16]:

# Find the intersection of rows and columns
common_rows = mcf7_smartS_raw.index.intersection(mcf7_smartS_filt.index)
common_columns = mcf7_smartS_raw.columns.intersection(mcf7_smartS_filt.columns)

# Create a new DataFrame based on the intersection of rows and columns
df_inters = mcf7_smartS_raw.loc[common_rows, common_columns]

# Compare the shapes
print(df_inters.shape == mcf7_smartS_filt.shape)


True


It is indeed true. Now we check if there are rows or columns with missing values in the raw dataset. We see that there are not. The same holds for rows or columns with just zeros.

In [7]:
# Drop rows with any missing values
df_ = mcf7_smartS_raw.dropna(axis=0, how='any')

# Drop columns with any missing values
df_ = df_.dropna(axis=1, how='any')


# Display the resulting dataset
print("Shape of df_:", df_.shape)




Shape of df_: (383, 22934)


In [10]:
# Drop columns that are all zeros
df_nonzero = mcf7_smartS_raw.loc[:, (mcf7_smartS_raw != 0).any(axis=0)]

# Drop rows that are all zeros
df_nonzero= df_nonzero.loc[(df_nonzero != 0).any(axis=1), :]


# Display the resulting dataset
print("Shape of df_nonzero:", df_nonzero.shape)




Shape of df_nonzero: (383, 22934)


Now we take a look at the removed rows and removed columns.

In [18]:

rows_removed = mcf7_smartS_raw.loc[~mcf7_smartS_raw.index.isin(mcf7_smartS_filt.index)]

# Step 2: Compare Columns
columns_removed = mcf7_smartS_raw.columns.difference(mcf7_smartS_filt.columns).tolist()


# Display Results
print("Rows Removed:", len(rows_removed))
print("Columns Removed:", len(columns_removed))
print("\nColumns Removed:")
print(columns_removed)  # Display the removed columns


Rows Removed: 70
Columns Removed: 3989

Columns Removed:
['A1CF', 'A2M', 'A2ML1', 'A2MP1', 'AANAT', 'ABCA13', 'ABCA4', 'ABCA8', 'ABCB11', 'ABCC13', 'ABCC5-AS1', 'ABCC9', 'ABCG4', 'ABHD12B', 'ABHD17AP4', 'ABHD17AP5', 'ABITRAMP1', 'ABO', 'ABRA', 'ABT1P1', 'ACAN', 'ACAP2-IT1', 'ACER1', 'ACER2P1', 'ACKR4P1', 'ACOXL-AS1', 'ACP4', 'ACSL5', 'ACSM4', 'ACTBL2', 'ACTL6B', 'ACTN2', 'ACTN3', 'ACTP1', 'ACTR3-AS1', 'ADAM21', 'ADAM21P1', 'ADAM29', 'ADAMTS10', 'ADAMTS14', 'ADAMTS15', 'ADAMTS2', 'ADAMTS20', 'ADAMTS3', 'ADAMTS4', 'ADAMTS5', 'ADAMTS6', 'ADAMTS7P1', 'ADARB2', 'ADCY10', 'ADCY2', 'ADCY4', 'ADGRA1', 'ADGRA1-AS1', 'ADGRD2', 'ADGRE1', 'ADGRE3', 'ADGRF5P1', 'ADGRG3', 'ADH4', 'ADH5P2', 'ADH6', 'ADIPOQ', 'ADIPOQ-AS1', 'ADM-DT', 'ADORA3', 'ADRA1B', 'ADRA1D', 'ADTRP', 'AFAP1L1', 'AFF2', 'AGBL1', 'AGBL4-AS1', 'AGBL5-AS1', 'AGGF1P1', 'AGK-DT', 'AGPAT4', 'AGRP', 'AGT', 'AGXT', 'AHCYP2', 'AHCYP4', 'AHCYP7', 'AHSG', 'AIFM3', 'AJAP1', 'AK4P6', 'AKAP14', 'AKIRIN1P2', 'AKIRIN2P1', 'AKR1B1', 'AKR1C4', 'AKR7

By looking at some of those columns in the text file, we see that there is a common trend. It seems that these columns contain lots of zero values for the corresponding gene expression in the observation. So we try to verify the hypotesis that the removal of a column is due to a vertical sum along the column that is below a certain threshold.

So we look at the smallest vertical su among the non-removed columns, and at the highest among the removed ones. If the first is higher than the second, our hypothesis is correct, else, we need to make another try. 

In [32]:
mcf7_smartS_raw[columns_removed].sum(axis=0).max()

1530

In [20]:
mcf7_smartS_filt.sum(axis=0).min()

6

As 6 < 1530, our hypothesis was incorrect. Now we take a look at the non-removed columns that have 6 as vertical sum.

In [21]:
# Get the names of the columns not removed where the sum of the column values is 6
columns_not_rem_with_sum_6 = mcf7_smartS_filt.columns[mcf7_smartS_filt.sum() == 6].tolist()

# Display the column names
print("Columns not rem where the sum is 6:")
print(columns_not_rem_with_sum_6)

Columns not rem where the sum is 6:
['MIR6726', 'DYNLL1P3', 'NUTF2P4', 'HNRNPA1P43', 'RNU6-307P', 'NSA2P1', 'GAPDHP25', 'ATP1B3P1', 'RPL21P37', 'SUPT4H1P1', 'FUNDC2P2', 'ACTR1AP1', 'NPM1P32', 'IDI1P1', 'UBBP3', 'HMGB1P4', 'HMGB1P9', 'RNU6-268P', 'RNU7-18P', 'HNRNPA1P67', 'RNU6-469P', 'RACK1P3', 'BTF3P13', 'RPS26P28', 'RPS4XP6', 'PDCD5P2', 'MIR1244-2', 'SUMO2P6', 'COX11P1', 'ATP5PBP6', 'KATNBL1P6', 'PSMC1P2', 'TUBG1P', 'PMS2P11', 'TFDP1P2', 'RPS11P7', 'SDCBPP3', 'AP1M2P1', 'MKRN5P', 'NT5DC1P2', 'RAC1P4', 'MAPK6P4', 'BRIX1P1', 'GLTPP1', 'EIF4A2P3', 'MIR6753', 'RBMXP3', 'DDX18P5', 'PSME2P6', 'SDHCP2', 'RPL15P13', 'MFAP5', 'PSMC1P8', 'NACAP8', 'MIR6762', 'HNRNPA3P5', 'ZDHHC20P4', 'RPL7P44', 'RANBP20P', 'RPL12P6', 'COX5AP1', 'EIF2S2P1', 'ELOCP2', 'FDPSP4', 'PPIAP46', 'RPL21P116', 'RRN3P1', 'NDUFA5P11', 'RPS27AP1', 'KRT31', 'RPL21P4', 'RPS26P8', 'RNF126P1', 'MIR6787', 'WBP11P1', 'LIVAR', 'MIR7108', 'CEACAM8', 'ZNF415', 'ZNF665', 'VDAC1P6', 'U2AF1L5', 'GAPDHP14', 'PSMD4P1', 'MIR4760']


As we did before, we take a look at some of them and we notice that each of these contains exactly six 1s and the rest are zeros. 
At the same time, we take a look at some of the removed ones with sum 6, and we notice that there are always 5 or less non-zero numbers that contribute to the sum.

In [22]:

# Identify Columns Removed
columns_removed = mcf7_smartS_raw.columns.difference(mcf7_smartS_filt.columns)

# Identify Columns Not Removed
columns_not_removed = mcf7_smartS_raw.columns.intersection(mcf7_smartS_filt.columns)

# Get the names of the columns not removed where the sum of the column values is 6
columns_not_removed_with_sum_6 = columns_not_removed[mcf7_smartS_raw[columns_not_removed].sum() == 6].tolist()
columns_removed_with_sum_6 = columns_removed[mcf7_smartS_raw[columns_removed].sum() == 6].tolist()

# Display the column names
print("Columns not removed where the sum is 6:")
print(columns_not_removed_with_sum_6)

print("Columns removed where the sum is 6:")
print(columns_removed_with_sum_6)

Columns not removed where the sum is 6:
['MIR6726', 'DYNLL1P3', 'NUTF2P4', 'HNRNPA1P43', 'RNU6-307P', 'ATP1B3P1', 'SUPT4H1P1', 'FUNDC2P2', 'ACTR1AP1', 'IDI1P1', 'HMGB1P4', 'HMGB1P9', 'RNU6-268P', 'RNU7-18P', 'HNRNPA1P67', 'RNU6-469P', 'BTF3P13', 'RPS26P28', 'RPS4XP6', 'PDCD5P2', 'MIR1244-2', 'COX11P1', 'ATP5PBP6', 'KATNBL1P6', 'PSMC1P2', 'TUBG1P', 'PMS2P11', 'TFDP1P2', 'RPS11P7', 'SDCBPP3', 'MKRN5P', 'NT5DC1P2', 'BRIX1P1', 'GLTPP1', 'EIF4A2P3', 'MIR6753', 'RBMXP3', 'DDX18P5', 'SDHCP2', 'RPL15P13', 'NACAP8', 'MIR6762', 'HNRNPA3P5', 'ZDHHC20P4', 'RPL7P44', 'RANBP20P', 'RPL12P6', 'COX5AP1', 'ELOCP2', 'FDPSP4', 'PPIAP46', 'RRN3P1', 'NDUFA5P11', 'RPS27AP1', 'RPL21P4', 'RNF126P1', 'MIR6787', 'WBP11P1', 'LIVAR', 'CEACAM8', 'ZNF415', 'VDAC1P6', 'U2AF1L5', 'GAPDHP14', 'MIR4760']
Columns removed where the sum is 6:
['ADCY10', 'ANKRD30A', 'ANKRD7', 'ATP2B2', 'ATP8A1-DT', 'B3GALT1-AS1', 'BCAP31P1', 'BCAR4', 'BLZF2P', 'C4B', 'C9orf57', 'CAMTA1-AS1', 'CCDC194', 'CCDC39-AS1', 'CCNYL4', 'CDCA4P4', 'CD

So this is our new hypotesis, if a column has 5 or less nonzero values, it gets deleted. We check if this is correct by looking at the intersection between the columns with more than 5 nonzero values in the raw dataset, and the columns of the filtered one. If the shape of this intersection matches the one of the latter, our hypotesis is correct.

In [23]:
# Count the number of non-zero values in each column in the raw dataset
non_zero_counts_per_column = (mcf7_smartS_raw != 0).sum(axis=0)

# Identify columns with more than 5 non-zero values
columns_with_more_than_5_non_zeros = non_zero_counts_per_column[non_zero_counts_per_column > 5].index

# Find the intersection of these columns with the columns of the filtered dataset
common_columns = columns_with_more_than_5_non_zeros.intersection(mcf7_smartS_filt.columns)

# Create a new DataFrame based on the intersection of columns
df_inters = mcf7_smartS_filt[common_columns]

# Compare the shapes
is_hypothesis_correct = df_inters.shape[1] == mcf7_smartS_filt.shape[1]

# Display the result
print("Is the hypothesis correct?", is_hypothesis_correct)
print("Shape of the filtered dataset:", mcf7_smartS_filt.shape)
print("Shape of the intersection dataset:", df_inters.shape)


Is the hypothesis correct? True
Shape of the filtered dataset: (313, 18945)
Shape of the intersection dataset: (313, 18945)


We see that every column with that characteristic was removed, because every column of the filtered dataset has that characteristic. But as we can see in the next cell code, there are still some columns that were removed but had more than 5 nonzero values.

In [31]:
# Count the number of non-zero values in each column
non_zero_counts = (mcf7_smartS_raw != 0).sum(axis=0)

# Count columns with more than 5 non-zero values
columns_with_more_than_5_non_zeros = (non_zero_counts > 5).sum()

# Display the result
print("Number of columns with more than 5 non-zero values:", columns_with_more_than_5_non_zeros)
print("Columns remaining to check:",columns_with_more_than_5_non_zeros - mcf7_smartS_filt.shape[1])

Number of columns with more than 5 non-zero values: 19182
Columns remaining to check: 237


There are indeed 237 columns that were removed for other reasons. We now print their indexes and take a look at them.

In [44]:

# Identify Columns Not Removed
columns_not_removed = mcf7_smartS_raw.columns.intersection(mcf7_smartS_filt.columns)

# Count the number of non-zero values in each column in the raw dataset
non_zero_counts = (mcf7_smartS_raw != 0).sum(axis=0)

# Identify columns with more than 5 non-zero values
columns_with_more_than_5_non_zeros = non_zero_counts[non_zero_counts > 5].index


# Find columns that meet both criteria
diff_df = columns_with_more_than_5_non_zeros.difference(columns_not_removed)

print(diff_df.shape)

np.set_printoptions(threshold=np.inf)

# Extract the sub-dataset containing these columns
sub_dataset = mcf7_smartS_raw[diff_df]

# Generate and display the description of the sub-dataset
sub_dataset_description = sub_dataset.describe()

# Display the resulting columns
print("Columns removed with more than 5 non-zero values:")
print(diff_df.to_numpy())

# Display the description of the sub-dataset
# print(sub_dataset_description)
#print(sub_dataset)


(237,)
Columns removed with more than 5 non-zero values:
['A2M' 'A2ML1' 'ABCC13' 'ABHD12B' 'ADIPOQ' 'ADM-DT' 'AJAP1' 'ALOX15P1'
 'AMDHD1' 'AMPH' 'ANGPT4' 'ANKLE1' 'AOC1' 'AQP8' 'ARAP3' 'ARG1'
 'ARHGEF18-AS1' 'ARL9' 'ASAP1-IT2' 'BACH1-AS1' 'BANK1' 'BATF2' 'BMS1P3'
 'C17orf50' 'C3orf35' 'CA5A' 'CABP1' 'CACNA1A' 'CALD1' 'CCBE1' 'CCN3'
 'CD70' 'CELF3' 'CELF4' 'CES3' 'CHMP4BP1' 'CHRNG' 'COLGALT2' 'CPB2-AS1'
 'CPLX3' 'CUX2' 'CYP2F1' 'DAND5' 'DDC' 'DDR1-DT' 'DDX10P2' 'DHX58'
 'DNAH12' 'DNAJC8P1' 'DPP3P1' 'ELMOD1' 'ENPP3' 'EPO' 'EPSTI1' 'EYS'
 'FAM135A' 'FAM71F2' 'FER1L6' 'FER1L6-AS2' 'FEZF1-AS1' 'FNDC8' 'FST'
 'FTH1P1' 'GAPDHP39' 'GOLGA8R' 'GOSR2-DT' 'GPR150' 'GZMM' 'H3P42' 'HHIP'
 'HLA-DQA2' 'HLA-S' 'HMGA1P3' 'HMGB1P26' 'HMGN2P23' 'HMX3' 'HNRNPA1P35'
 'HNRNPA3P9' 'HOXB-AS1' 'HSPB6' 'HSPE1P6' 'HSPE1P7' 'HTR2B' 'IGF2-AS'
 'IGLL3P' 'IGLV1-51' 'IPCEF1' 'IQGAP2' 'ITGA1' 'ITGB7' 'ITIH4' 'KCNC1'
 'KIF26B' 'KLF1' 'KLHL6' 'KLRG1' 'KRT37' 'LARP1P1' 'LCN2' 'LINC00310'
 'LINC00572' 'LINC00661' 'LINC0096

By looking at some of them, we noticed that they have very sparse columns. Since the filtered dataset had 70 rows less than the raw one, it is possible that the procedure explained before was performed after the removal of the rows. Hence that the removal of the rows had happend before the one of the columns. The following code checks that

In [48]:

# Identify rows that were removed
removed_rows = mcf7_smartS_raw.index.difference(mcf7_smartS_filt.index)

# Create a new version of the raw dataset with those rows removed
raw_without_removed_rows = mcf7_smartS_raw.drop(index=removed_rows)

# Count the number of non-zero values in each column of the new dataset
non_zero_counts_after_row_removal = (raw_without_removed_rows != 0).sum(axis=0)

# Identify columns that had more than 5 non-zero values in the original raw dataset
non_zero_counts_original = (mcf7_smartS_raw != 0).sum(axis=0)
columns_with_more_than_5_non_zeros_original = non_zero_counts_original[non_zero_counts_original > 5].index

# Identify columns with more than 5 non-zero values in the new dataset (after row removal)
columns_with_more_than_5_non_zeros_after_row_removal = non_zero_counts_after_row_removal[non_zero_counts_after_row_removal > 5].index

# Find columns that were removed but had more than 5 non-zero values in the original dataset
diff_df = columns_with_more_than_5_non_zeros_original.difference(mcf7_smartS_filt.columns)

# Check which of these columns have less than or equal to 5 non-zero values after row removal
columns_with_less_or_equal_5_after_row_removal = non_zero_counts_after_row_removal[diff_df]
columns_removed_due_to_row_removal = columns_with_less_or_equal_5_after_row_removal[columns_with_less_or_equal_5_after_row_removal <= 5].index

# Display the result
print("Columns that were removed and had more than 5 non-zero values originally:")
print(diff_df.to_numpy())

print("Columns from above that have less than or equal to 5 non-zero values after row removal:")
print(columns_removed_due_to_row_removal.to_numpy())

# Validate hypothesis
hypothesis_correct = (diff_df.to_numpy().tolist() == columns_removed_due_to_row_removal.to_numpy().tolist())
print("Is the hypothesis correct?", hypothesis_correct)


Columns that were removed and had more than 5 non-zero values originally:
['A2M' 'A2ML1' 'ABCC13' 'ABHD12B' 'ADIPOQ' 'ADM-DT' 'AJAP1' 'ALOX15P1'
 'AMDHD1' 'AMPH' 'ANGPT4' 'ANKLE1' 'AOC1' 'AQP8' 'ARAP3' 'ARG1'
 'ARHGEF18-AS1' 'ARL9' 'ASAP1-IT2' 'BACH1-AS1' 'BANK1' 'BATF2' 'BMS1P3'
 'C17orf50' 'C3orf35' 'CA5A' 'CABP1' 'CACNA1A' 'CALD1' 'CCBE1' 'CCN3'
 'CD70' 'CELF3' 'CELF4' 'CES3' 'CHMP4BP1' 'CHRNG' 'COLGALT2' 'CPB2-AS1'
 'CPLX3' 'CUX2' 'CYP2F1' 'DAND5' 'DDC' 'DDR1-DT' 'DDX10P2' 'DHX58'
 'DNAH12' 'DNAJC8P1' 'DPP3P1' 'ELMOD1' 'ENPP3' 'EPO' 'EPSTI1' 'EYS'
 'FAM135A' 'FAM71F2' 'FER1L6' 'FER1L6-AS2' 'FEZF1-AS1' 'FNDC8' 'FST'
 'FTH1P1' 'GAPDHP39' 'GOLGA8R' 'GOSR2-DT' 'GPR150' 'GZMM' 'H3P42' 'HHIP'
 'HLA-DQA2' 'HLA-S' 'HMGA1P3' 'HMGB1P26' 'HMGN2P23' 'HMX3' 'HNRNPA1P35'
 'HNRNPA3P9' 'HOXB-AS1' 'HSPB6' 'HSPE1P6' 'HSPE1P7' 'HTR2B' 'IGF2-AS'
 'IGLL3P' 'IGLV1-51' 'IPCEF1' 'IQGAP2' 'ITGA1' 'ITGB7' 'ITIH4' 'KCNC1'
 'KIF26B' 'KLF1' 'KLHL6' 'KLRG1' 'KRT37' 'LARP1P1' 'LCN2' 'LINC00310'
 'LINC00572' 'LIN

As we can see, this is the case. Now we need to understand how were those 70 rows removed.

In [49]:
# Identify rows that were removed
removed_rows = mcf7_smartS_raw.index.difference(mcf7_smartS_filt.index)

# Print the indexes of the removed rows
print("Indexes of the removed rows:")
print(removed_rows.to_numpy())


Indexes of the removed rows:
['output.STAR.1_A11_Hypo_S29_Aligned.sortedByCoord.out.bam'
 'output.STAR.1_A1_Norm_S1_Aligned.sortedByCoord.out.bam'
 'output.STAR.1_A7_Hypo_S25_Aligned.sortedByCoord.out.bam'
 'output.STAR.1_A8_Hypo_S26_Aligned.sortedByCoord.out.bam'
 'output.STAR.1_A9_Hypo_S27_Aligned.sortedByCoord.out.bam'
 'output.STAR.1_B10_Hypo_S76_Aligned.sortedByCoord.out.bam'
 'output.STAR.1_B1_Norm_S49_Aligned.sortedByCoord.out.bam'
 'output.STAR.1_B2_Norm_S50_Aligned.sortedByCoord.out.bam'
 'output.STAR.1_B3_Norm_S51_Aligned.sortedByCoord.out.bam'
 'output.STAR.1_B9_Hypo_S75_Aligned.sortedByCoord.out.bam'
 'output.STAR.1_C3_Norm_S99_Aligned.sortedByCoord.out.bam'
 'output.STAR.1_C7_Hypo_S121_Aligned.sortedByCoord.out.bam'
 'output.STAR.1_C8_Hypo_S122_Aligned.sortedByCoord.out.bam'
 'output.STAR.1_C9_Hypo_S123_Aligned.sortedByCoord.out.bam'
 'output.STAR.1_D10_Hypo_S172_Aligned.sortedByCoord.out.bam'
 'output.STAR.1_D12_Hypo_S174_Aligned.sortedByCoord.out.bam'
 'output.STAR.1_D1_

In [50]:

# Identify rows that were removed
removed_rows = mcf7_smartS_raw.index.difference(mcf7_smartS_filt.index)

# Select the removed rows from the raw dataset
removed_rows_data = mcf7_smartS_raw.loc[removed_rows]

# Calculate the sum of inputs in each row for the removed rows
removed_rows_sum = removed_rows_data.sum(axis=1)

# Print the sums of the removed rows
print("Sum of inputs in each row for the removed rows:")
print(removed_rows_sum.to_numpy())


Sum of inputs in each row for the removed rows:
[    281   23499   51389   22704   87386 2287165  131367  128171  208267
     129  160882  103751  127574    9790  239896  564282  152337  170166
   77094    7366      19   73412     242   45972  110173   94119   69725
  106668  180121     157   96688   98198   38467   28803   76280      13
    4701  174355   78657  114469     887   22769  122182       1   14645
   99079      22    2267    6013 2069263     113 2162257 2308057    6201
 1196909      36   36673  871703    1519    8773     153     422      30
  595098  817482    8941   70269 1302612    6572  795636]


In [51]:


# Calculate the sum of inputs in each row for the filtered dataset
filtered_rows_sum = mcf7_smartS_filt.sum(axis=1)

# Print the sums of the rows present in the filtered dataset
print("Sum of inputs in each row for the filtered dataset:")
print(filtered_rows_sum.to_numpy())


Sum of inputs in each row for the filtered dataset:
[ 936059 1982038  333237 1289120 1728927 1439437 1545461 1616084  940724
  310372 1559136 1298616  388450 1106482  776727 1427685 1533968  293995
  534204  285546 1521105 1429833  943291  461056 1099121  760684  451975
  761218 1416180  263343  773248  266752 1375304  268894  644286 1229842
 1775947  819236  470831  374756  805460 1393651 1244028 1505568  382918
 1265938  659712 1822649 1265845 1801662 1606844 1590490 1164751  923967
 1385844 1509423 1563203 1916212 1354593 1853247 1852260 1190040 1085533
 1604592 1550826 1717389 1090990  985759 1043203 1734426 1840082 1374320
 1639861 1770634 1580080 1473480 1666952 1352802 1470531  957652 1814330
 1666022 1414220 1255645 1547154 1625442  432363 1045592 1618615 1724182
 1970851  334200  318188 1559445 1504638 1472393 1469857 1665543 1904769
 1488396  897753  973800 1617331 1365857 1570662 1679923 1605773 1610803
 1661102 1361039 1391606 1446626 1232708  862168  604070  597671 1754394

In [52]:
# Identify rows that were removed
removed_rows = mcf7_smartS_raw.index.difference(mcf7_smartS_filt.index)

# Select the removed rows from the raw dataset
removed_rows_data = mcf7_smartS_raw.loc[removed_rows]

# Compute the number of nonzero values in each row for the removed rows
removed_rows_nonzero_count = (removed_rows_data != 0).sum(axis=1)

# Print the number of nonzero values for each row in the removed ones
print("Number of nonzero values for each row in the removed ones:")
print(removed_rows_nonzero_count.to_numpy())


Number of nonzero values for each row in the removed ones:
[  188  4739  6549  4594  5945  9307  8829  8571  8153    58  9010  7997
  7856  3105  8567  2778  8614  8956  7757  3166     7  6379   199  6663
  8022  8390  6242  7438  7573   132  8143  8090  5148  5013  6625    13
  2200  8035  6863  8380   574  4617  7389     1  3976  7598    22  1120
  2195  8362    86  6745  6932  2620 12519    33  5260 12504  1025  3321
    98    97    27  2636  3953  2979  3744  9619  3577  1975]


In [53]:


# Compute the number of nonzero values in each row for the non-removed rows
non_removed_rows_nonzero_count = (mcf7_smartS_filt != 0).sum(axis=1)

# Print the number of nonzero values for each row in the non-removed ones
print("Number of nonzero values for each row in the non-removed ones:")
print(non_removed_rows_nonzero_count.to_numpy())


Number of nonzero values for each row in the non-removed ones:
[ 8664  6433 10007  9778 10598 11663 10584  9623 10546  8863 10892 11254
  9066 10324  7986  9791 10020  9846  9571  8436  9349 10597  9717 10436
 10031  9516  7478 10085 10648  9533  9758  8161  8872  8855 10131 11050
 10709 10160  7938  9073  7768  9971  8940  8900  9810 10663  9029  8905
  9586  7481 10478 10641 11779 11243 11027 10919  9796  8277  7017  9014
  8257 11294 11780 10282 10686 10124 12217  9689 10303  9451  8936 11108
 11910 10425 10995 11196 11040 10748 10840 10953 10629 10208 11432 11822
 10539  9661  9136  8308  9605 10067  8471  9218  9601 11042  9818 11102
 10486  8916  8869 10231 11052 10698  5527 11868 10391 10889 10740 10093
  9980 10203 10384  8963 10253  8199  9866 10617 10209  9224 11420 11125
 11918  8432  9865  8482  9780 10697 10015  9851 10339 10890 11805 11342
  6416 10738  9165  9978  8959  8846 10392 11060 10761 11985  9515 10650
 10264  7807 10150  7051  5974 11312 11407 10598 10711 10519 

In [94]:

# Count the number of non-zero values in each row
non_zero_counts_per_row = (mcf7_smartS_raw != 0).sum(axis=1)

# Identify rows with strictly less than 6 non-zero values
rows_with_less_than_6_non_zeros = mcf7_smartS_raw[non_zero_counts_per_row < 7895]

# Display the result
print("Number of rows with strictly less than 6 non-zero values:", rows_with_less_than_6_non_zeros.shape[0])
print("First few rows with strictly less than 6 non-zero values:")
#print(rows_with_less_than_6_non_zeros.head())


Number of rows with strictly less than 6 non-zero values: 70
First few rows with strictly less than 6 non-zero values:


In [110]:

# Identify Rows Removed
rows_removed = mcf7_smartS_raw.loc[~mcf7_smartS_raw.index.isin(mcf7_smartS_filt.index)]

# Count the number of non-zero values in each row in the raw dataset
non_zero_counts_per_row = (mcf7_smartS_raw != 0).sum(axis=1)

# Identify rows with strictly less than 6 non-zero values
rows_with_less_than_6_non_zeros = mcf7_smartS_raw[non_zero_counts_per_row < 7895]

print(len(rows_with_less_than_6_non_zeros))

# Find the intersection of these rows with the rows removed
intersection_rows = rows_with_less_than_6_non_zeros.index.intersection(rows_removed.index)

# Display the result
print("Number of rows with strictly less than 6 non-zero values that were also removed:", len(intersection_rows))
print("First few rows in the intersection:")
print(mcf7_smartS_raw.loc[intersection_rows].head())


70
Number of rows with strictly less than 6 non-zero values that were also removed: 51
First few rows in the intersection:
                                                    WASH7P  MIR6859-1  WASH9P  \
output.STAR.1_A11_Hypo_S29_Aligned.sortedByCoor...       0          0       0   
output.STAR.1_A1_Norm_S1_Aligned.sortedByCoord....       0          0       0   
output.STAR.1_A7_Hypo_S25_Aligned.sortedByCoord...       0          0       0   
output.STAR.1_A8_Hypo_S26_Aligned.sortedByCoord...       0          0       0   
output.STAR.1_A9_Hypo_S27_Aligned.sortedByCoord...       0          0       0   

                                                    OR4F29  MTND1P23  \
output.STAR.1_A11_Hypo_S29_Aligned.sortedByCoor...       0         0   
output.STAR.1_A1_Norm_S1_Aligned.sortedByCoord....       0         0   
output.STAR.1_A7_Hypo_S25_Aligned.sortedByCoord...       0         0   
output.STAR.1_A8_Hypo_S26_Aligned.sortedByCoord...       0         0   
output.STAR.1_A9_Hypo_S27_Alig

In [106]:
non_zero_counts_per_row = (mcf7_smartS_raw != 0).sum(axis=1)
pierino = (mcf7_smartS_raw).sum(axis=1)

pierino.describe()


count      383.000000
mean      9124.219321
std       2693.309249
min          1.000000
25%       8456.500000
50%       9907.000000
75%      10789.000000
max      12519.000000
dtype: float64

In [107]:
non_zero_counts_per_row.describe()

count      383.000000
mean      9124.219321
std       2693.309249
min          1.000000
25%       8456.500000
50%       9907.000000
75%      10789.000000
max      12519.000000
dtype: float64